scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1122 @@
+            elif variant == "ia":
+                self._check_errorcode(unur_tdr_set_variant_ia(self.par))
+            elif variant == "gw":
+                self._check_errorcode(unur_tdr_set_variant_gw(self.par))
+
+            self._check_errorcode(unur_tdr_set_usedars(self.par, use_dars))
+            self._check_errorcode(unur_tdr_set_max_sqhratio(self.par, max_sqhratio))
+            self._check_errorcode(unur_tdr_set_max_intervals(self.par, max_intervals))
+            self._check_errorcode(unur_tdr_set_usecenter(self.par, use_center))
+            self._check_errorcode(unur_tdr_set_usemode(self.par, use_mode))
+            self._check_errorcode(unur_tdr_set_guidefactor(self.par, guide_factor))
+
+            self._set_rng(seed)
+        finally:
+            _lock.release()
+            release_unuran_callback(&callback)
+
+    cdef object _validate_args(self, params, domain, c, cpoints, variant):
+        domain = _validate_domain(domain)
+        if np.isnan(c):
+            raise ValueError("`c` must be a non-nan value.")
+        if variant not in {"ps", "gw", "ia"}:
+            raise ValueError("Invalid option for the `variant`.")
+        if not np.isscalar(cpoints):
+            self.cpoints_array = np.ascontiguousarray(cpoints, dtype=np.float64)
+            if len(self.cpoints_array) == 0:
+                raise ValueError("`cpoints` must either be a scalar or a "
+                                 "non-empty array.")
+        else:
+            self.cpoints_array = None
+
+        params = tuple(params)
+
+        return params, domain, c, cpoints, variant
+
+    @property
+    def sqhratio(self):
+        """
+        Get the current ratio (area below squeeze) / (area below hat) for the
+        generator.
+        """
+        return unur_tdr_get_sqhratio(self.rng)
+
+    @property
+    def hat_area(self):
+        """Get the area below the hat for the generator."""
+        return unur_tdr_get_hatarea(self.rng)
+
+    @property
+    def squeeze_area(self):
+        """Get the area below the squeeze for the generator."""
+        return unur_tdr_get_squeezearea(self.rng)
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    cdef inline void _ppf_hat(self, const double *u, double *out, size_t N) except *:
+        cdef:
+            ccallback_t callback
+            size_t i
+        init_unuran_callback(&callback, self._callback_wrapper, self.params)
+        _lock.acquire()
+        try:
+            self._messages.clear()
+            unur_set_stream(self._messages.handle)
+            for i in range(N):
+                out[i] = unur_tdr_eval_invcdfhat(self.rng, u[i], NULL, NULL, NULL)
+            msg = self._messages.get()
+            if msg:
+                raise UNURANError(msg)
+        finally:
+            _lock.release()
+            release_unuran_callback(&callback)
+
+    def ppf_hat(self, u):
+        """
+        Evaluate the inverse of the CDF of the hat distribution at `u`.
```

Comment:
Is this useful? 

Without knowing the method very well, I'm imagining that the hat CDF is the CDF of the "proposal distribution" (in the language of [Wikipedia](https://en.wikipedia.org/wiki/Rejection_sampling)). Unless the proposal distribution very closely approximates to the true PDF, I don't understand how this would be useful.

Perhaps this question is super ignorant and I need to read more about this method, e.g. the tutorial.

Reply:
> Is this useful?

Might be.

> Unless the proposal distribution very closely approximates to the true PDF, I don't understand how this would be useful

Generally, `ppf_hat` matches the exact `ppf` up to 1 or 2 decimal places. The accuracy of `ppf_hat` can be controlled by setting the `max_intervals` and `max_sqhratio` parameters. By default `max_sqhratio` is equal to `0.99` which means that UNU.RAN will choose the intervals such that the ratio of area under squeeze and area under hat will be above `0.99`. If we change this to something like `0.99999`, we can get a more accurate estimate of the PPF (typically maximum up to 5 decimal places from my experiments and tests). This comes at an expense of more PDF evaluations and high `max_intervals` value (around `10000`) during setup.

For example, [see this test](https://github.com/scipy/scipy/blob/246febbc7a9621b6e6752053d266b7ff7a9d7af0/scipy/stats/tests/test_sampling.py#L445):

```python

```

> Perhaps this question is super ignorant and I need to read more about this method, e.g. the tutorial.

I haven't explained in the tutorial how to get a better estimate of the PPF. I will add a description with an example.

Reply:
Hmm. Well, throwing this out there, seeking unity in the API: if the reason it might be useful is to approximate the PPF, could it just be called `ppf` and described as an approximation of the `ppf` with some additional notes about it actually being the "hat distribution" and thoughts about accuracy?

Reply:
> (Although I understand that the accuracy of the approximation in `NumericalInverseHermite` has to be much better than this PPF-hat in order to get accurate random samples, so I can also see the argument in making a distinction.)

That's exactly why I don't want to name it `ppf`. Users might misunderstand this as being an accurate approximation or exact PPF. UNU.RAN also makes this distinction between accurate and inaccurate approximation.

Reply:
I would also keep the name `ppf_hat`. I don't think it is very relevant for most users but it is nice to have: usually, you can easily compare the original pdf and the pdf of the hat. but computing the ppf of the hat is some extra work and it gives you a better idea of how close the hat is than looking at the pdf.

Reply:
OK. I'll resolve this. But if this is intended to be a "power user" thing rather than core functionality, consider not exposing it until requested. The documentation and intended purpose just doesn't seem as complete to me as the rest of the class.